In [1]:
import pandas as pd 
from sklearn.model_selection import train_test_split


In [13]:
#Here from this data our aim is to find whether the audiobook will be purchased or not in future
#that is why we choose logistic regression first.
# we can set the data for model training now.

scaled_data=pd.read_csv("audioboook_scaled.csv")

# now we can define features and target

X=scaled_data.drop(columns=['ID','Target'])
Y=scaled_data['Target']


In [21]:
#now it is our time to train test ans split data
#with the X_train and y_train we finds a model that actually predicts the purchase or not.
#then with the y_train is the new unseen data , which tries to predict whethere purchase will be done or not 
#then validates it with the actual y_test data.
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(X,Y,test_size=0.2,random_state=42,stratify=Y)

test_size=0.2
→ 20% of the data goes into the test set, and 80% into the training set.
(If you had 10,000 records → 8,000 train + 2,000 test).

random_state=42
→ ensures reproducibility.
(So if you run the code again, you’ll get the same split every time).
If you don’t set this, every run may give a different random split.

stratify=Y
→ very important in classification.
Ensures the same proportion of classes (0 and 1) is kept in both training and test sets.
Example: If your dataset has 70% non-purchasers (0) and 30% purchasers (1),
then both train and test will preserve that ratio (not distort it).

In [26]:
print("Shape of train dataset :",x_train.shape,y_train.shape)
print("Shape of test dataset :",x_test.shape,y_test.shape)

Shape of train dataset : (11266, 11) (11266,)
Shape of test dataset : (2817, 11) (2817,)


In [38]:
from sklearn.linear_model import LogisticRegression
#Initialize the model
log_reg=LogisticRegression(max_iter=1000,random_state=42)
log_reg.fit(x_train,y_train)

LogisticRegression(max_iter=1000, random_state=42)

In [40]:
#Predictions
y_pred=log_reg.predict(x_test)
y_prob=log_reg.predict_proba(x_test)[:,1]

In [56]:
#now we can go for the evaluation.
from sklearn.metrics import (accuracy_score,precision_score,recall_score,f1_score,
classification_report,confusion_matrix,roc_auc_score)

print("Accuracy :",accuracy_score(y_test,y_pred))
print("Precision_score :",precision_score(y_test,y_pred))
print("Recall score :",recall_score(y_test,y_pred))
print("f1 score :",f1_score(y_test,y_pred))

Accuracy : 0.9023784167554135
Precision_score : 0.9343434343434344
Recall score : 0.41387024608501116
f1 score : 0.5736434108527132


Accuracy (90.2%) → The model is correct 90% of the time. Sounds great, but accuracy alone can be misleading if the classes are imbalanced.

Precision (93.4%) → When the model predicts “purchase = 1”, it’s correct 93% of the time. This is very high, meaning false positives are low.

Recall (41.3%) → Out of all the actual purchases, the model only catches 41%.This is quite low — meaning it’s missing a lot of true buyers (high false negatives).

F1 score (57.3%) → This is the balance between precision and recall. Since recall is low, the F1 score drops.

In [69]:
#now we can use ROC_AUC here 
print("ROC AUC :",roc_auc_score(y_test,y_pred))
print("Classification Report :",classification_report(y_test,y_pred))
print("Confusion matrix :",confusion_matrix(y_test,y_pred))

ROC AUC : 0.7041925070087504
Classification Report :               precision    recall  f1-score   support

           0       0.90      0.99      0.94      2370
           1       0.93      0.41      0.57       447

    accuracy                           0.90      2817
   macro avg       0.92      0.70      0.76      2817
weighted avg       0.91      0.90      0.89      2817

Confusion matrix : [[2357   13]
 [ 262  185]]


Model is excellent at identifying non-buyers (99% recall for class 0).

Model struggles to catch buyers (41% recall).

This imbalance is common in real-world business datasets (class imbalance).

Precision (class 1 = buyers = 0.93): If the model predicts "buyer," it is correct 93% of the time. 

Recall (class 1 = buyers = 0.41): But it only finds 41% of actual buyers meaning many buyers are missed.

F1-score (class 1 = 0.57): A balance of precision & recall; low because recall drags it down.

Overall: Model is biased towards predicting non-buyers (class 0) accuracy looks good, but recall for buyers is weak.